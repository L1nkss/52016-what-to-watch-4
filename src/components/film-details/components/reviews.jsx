const renderReview = (reviews) => {
  return reviews.map((review) => {
    return (
      <div className="review" key={review.author}>
        <blockquote className="review__quote">
          <p className="review__text">{review.text}</p>

          <footer className="review__details">
            <cite className="review__author">{review.author}</cite>
            <time className="review__date" dateTime="2015-11-18">{review.date}</time>
          </footer>
        </blockquote>

        <div className="review__rating">{review.rating}</div>
      </div>
    );
  });
};


const Reviews = (props) => {
  return (
    <div className="movie-card__reviews movie-card__row">
      <div className="movie-card__reviews-col">
        {renderReview(props.data)}
      </div>
    </div>
  );
};

Reviews.propTypes = {
  data: propTypes.arrayOf(propTypes.object)
};

export default Reviews;
