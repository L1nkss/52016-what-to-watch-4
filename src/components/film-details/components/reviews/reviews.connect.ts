import Reviews from "@components/film-details/components/reviews/reviews";
import {connect} from "react-redux";

const mapStateToProps = (state) => ({
  reviews: state.REVIEWS.reviews,
  loading: state.REVIEWS.loading
});

export {Reviews};
export default connect(mapStateToProps, null)(Reviews);
