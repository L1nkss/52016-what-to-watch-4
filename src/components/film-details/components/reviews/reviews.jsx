export default class Reviews extends React.Component {
  constructor(props) {
    super(props);
    this.props.getReviews();
  }
  renderReview(reviews) {
    return reviews.map((review) => {
      return (
        <div className="review" key={review.id}>
          <blockquote className="review__quote">
            <p className="review__text">{review.comment}</p>
            <footer className="review__details">
              <cite className="review__author">{review.user.name}</cite>
              <time className="review__date" dateTime="2016-12-24">{review.date}</time>
            </footer>
          </blockquote>

          <div className="review__rating">{review.rating}</div>
        </div>
      );
    });
  }
  render() {
    if (this.props.loading || this.props.reviews === null) {
      return <div className="loader" />;
    }
    const halfReviews = this.props.reviews.splice(0, this.props.reviews.length / 2);
    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {this.renderReview(this.props.reviews)}
        </div>
        <div className="movie-card__reviews-col">
          {this.renderReview(halfReviews)}
        </div>
      </div>
    );
  }
}

Reviews.propTypes = {
  loading: propTypes.bool.isRequired,
  reviews: propTypes.array.isRequired,
  getReviews: propTypes.func.isRequired
};
