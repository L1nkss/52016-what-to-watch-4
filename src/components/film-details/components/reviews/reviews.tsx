import moment from "moment";

type TReview = {
  rating: number,
  comment: string
}

interface IReviews {
  loading: boolean,
  reviews: Array<TReview>
  getReviews: () => void
}

export default class Reviews extends React.Component<IReviews> {
  constructor(props) {
    super(props);
    // Перенести из констуктора
    this.props.getReviews();
  }
  renderReview(reviews) {
    return reviews.map((review) => {
      return (
        <div className="review" key={review.id}>
          <blockquote className="review__quote">
            <p className="review__text">{review.comment}</p>
            <footer className="review__details">
              <cite className="review__author">{review.user.name}</cite>
              <time
                className="review__date"
                dateTime={moment(review.date).format(`YYYY-M-D`)}>
                {moment(review.date).format(`MMMM D, YYYY`)}
              </time>
            </footer>
          </blockquote>

          <div className="review__rating">{review.rating}</div>
        </div>
      );
    });
  }
  render() {
    if (this.props.loading || this.props.reviews === null) {
      return <div className="loader" />;
    }
    const halfReviews = this.props.reviews.splice(0, this.props.reviews.length / 2);
    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">
          {this.renderReview(this.props.reviews)}
        </div>
        <div className="movie-card__reviews-col">
          {this.renderReview(halfReviews)}
        </div>
      </div>
    );
  }
}
