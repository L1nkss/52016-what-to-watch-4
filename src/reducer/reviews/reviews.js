import ActionType from "./constants/constants.js";
import ActionCreator from "./actions/actions";

const initialState = {
  reviews: null,
  loading: false,
  error: false
};

const Operation = {
  loadReviews: (id) => (dispatch, getState, api) => {
    dispatch(ActionCreator.loadFilmReviewsRequest());
    return api.getReviews(id)
      .then((response) => {
        dispatch(ActionCreator.loadFilmReviewsSuccess(response.data));
      })
      .catch((err) => {
        dispatch(ActionCreator.loadFilmReviewsError());
        throw err;
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_REVIEWS_REQUEST:
      return Object.assign({}, state, {loading: true, error: false});
    case ActionType.LOAD_REVIEWS_SUCCESS:
      return Object.assign({}, state, {loading: false, reviews: action.payload});
    case ActionType.LOAD_REVIEWS_ERROR:
      return Object.assign({}, state, {loading: false, error: true});
    default:
      return state;
  }
};

export {reducer, Operation};
