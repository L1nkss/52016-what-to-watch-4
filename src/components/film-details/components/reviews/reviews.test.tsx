import * as React from "react";
import * as renderer from "react-test-renderer";
import {Reviews} from "./reviews.connect";

const reviews = [
  {
    rating: 6,
    id: 5,
    comment: `Text`,
    user: {
      name: `Maxim`
    }
  },
  {
    id: 10,
    rating: 10,
    comment: `Text new`,
    user: {
      name: `Maxim`
    }
  }
];

describe(`Testing reviews component`, () => {
  it(`Component should successfully rendered`, () => {
    const tree = renderer.create(<Reviews loading={false} reviews={reviews} getReviews={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  })
});
