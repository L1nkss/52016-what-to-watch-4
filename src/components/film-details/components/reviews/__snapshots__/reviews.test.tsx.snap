// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing reviews component Component should successfully rendered 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Text new
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Maxim
          </cite>
          <time
            className="review__date"
            dateTime="2020-7-29"
          >
            July 29, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        10
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Text
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Maxim
          </cite>
          <time
            className="review__date"
            dateTime="2020-7-29"
          >
            July 29, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        6
      </div>
    </div>
  </div>
</div>
`;
